x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/DocumentPermissions.java

```diff
@@ -204,4 +250,37 @@ public String toString() {
         return "DocumentPermissions [queries=" + queries + ", scopedByQueries=" + limitedByQueries + "]";
     }
 
+    @Override
+    public void buildCacheKey(StreamOutput out) throws IOException {
+        assert false == (queries == null && limitedByQueries == null) : "one of queries and limited-by queries must be non-null";
+        if (queries != null) {
+            assert evaluatedQueries != null : "queries are not evaluated";
```

Comment:
This feels wrong. It means that it's only possible to call this method if you have called some other method first.

But in fact there's no reason for the caller to know that they should do that, or to even expect them to be linked.

It might be what we need to do right now, but I'm very reluctant to leave it this way.

The obvious answers are to either:
1. Inject enough services into the constructor (or a builder) to evaluate the queries immediately
2. Split this class into 2, so that there's a separate `EvaluatedDocumentPermissions` class that implements `CacheKey`.
 
I don't know if either of those are feasible.

Reply:
It's certainly non-ideal. But it is not really feasible to go with either of your options either. I thought about first one but it's cascading changes all over the place and it is questionable to inject `ScriptService` everywhere. The second option does not completely get rid of the "your must call some other method first" problem. It just moves it to a different place. 

I had another idea initially but didn't go with it. But maybe it is a better approach after all: We can modify the `buildCacheKey` signature to take `User` and `ScriptService`. They are not really needed at code level in most places and entirely not needed in practice, that's why I didn't go with it because they seem to be superfluous and I was considering the problem you raised to be less of an issue (e.g. some object needs to be `init` after instantiation to be really usable). It's a balancing act, right now I am OK with either one.

Reply:
As discussed, we agree to change the signature of `buildCacheKey` to pass in a `DlsQueryEvaluationContext`. Overall this change is the least intrusive.